### PR TITLE
impl(pubsub): add end modack event for exactly once

### PR DESCRIPTION
**Pull request by alevenberg**
[#13726](https://github.com/googleapis/google-cloud-cpp/issues/13726)

This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13920)

### CODE REVIEW

File: google/cloud/pubsub/internal/streaming_subscription_batch_source.cc [hunk 1]
```diff
@@ -222,7 +222,15 @@ void StreamingSubscriptionBatchSource::ExtendLeases(
   auto split = SplitModifyAckDeadline(std::move(request), kMaxAckIdsPerMessage);
   if (exactly_once_delivery_enabled_.value_or(false)) {
     lk.unlock();
-    for (auto& r : split) (void)ExtendLeasesWithRetry(stub_, cq_, std::move(r));
+    for (auto& r : split)
```
Comment (dbolduc) [inline google/cloud/pubsub/internal/streaming_subscription_batch_source.cc:225]: nit: `{` `}`

File: google/cloud/pubsub/internal/streaming_subscription_batch_source_test.cc [hunk 2]
```diff
@@ -1130,7 +1130,7 @@ TEST(StreamingSubscriptionBatchSourceTest, ExtendLeasesWithRetry) {
       std::make_shared<pubsub_testing::MockBatchCallback>();
   EXPECT_CALL(*mock_batch_callback, callback).Times(1);
   EXPECT_CALL(*mock_batch_callback, ModackStart).Times(2);
-  EXPECT_CALL(*mock_batch_callback, ModackEnd).Times(0);
+  EXPECT_CALL(*mock_batch_callback, ModackEnd).Times(2);
```
Comment (dbolduc) [inline google/cloud/pubsub/internal/streaming_subscription_batch_source_test.cc:1133]: I'd definitely prefer to have separate expectations for each ack id. Calling `ModackEnd("fake-001")` twice is not the same thing as calling `ModackEnd("fake-001")` once and `ModackEnd("fake-002")` once. But both would pass the expectations, as is.

```cc
EXPECT_CALL(*mock_batch_callback, ModackEnd("fake-001")); 
EXPECT_CALL(*mock_batch_callback, ModackEnd("fake-002"));
```

same goes for `ModackStart`

**Review by dbolduc — APPROVED** [review]

File: google/cloud/pubsub/internal/streaming_subscription_batch_source_test.cc (hunk 2 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/streaming_subscription_batch_source_test.cc]: Done

File: google/cloud/pubsub/internal/streaming_subscription_batch_source.cc (hunk 1 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/streaming_subscription_batch_source.cc]: Done

**Pull request merged**